### PR TITLE
Extract book into a submodule

### PR DESCRIPTION
**Pull request by steveklabnik**
Part of https://github.com/rust-lang/rust/issues/39588

We probably don't want to land this till after the beta branches on friday, but would still ❤️ a review from @alexcrichton , since I am a rustbuild noob.

This pr:

1. removes the book
2. adds it back in as a submodule
3. the submodule includes both the old book and the new book
4. it also includes an index page explaining the difference in editions
5. it also includes redirect pages for the old book URLs.
6. so we build all that stuff too.

r? @alexcrichton

### CODE REVIEW

**Comment by steveklabnik** [comment]
I pushed two more commits; one fixes a trailing whitespace issue, the last one exempts the book from tidy. I _think_ we don't actually want that, but for now it will let things build.

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -51,6 +51,82 @@ pub fn rustbook(build: &Build, target: &str, name: &str) {
...
+
+    let version_input = build.src.join("src/doc/version_info.html.template");
+    let version_info = out.join("version_info.html");
+
+    if !up_to_date(&version_input, &version_info) {
+        let mut info = String::new();
+        t!(t!(File::open(&version_input)).read_to_string(&mut info));
+        let info = info.replace("VERSION", &build.rust_release())
+                       .replace("SHORT_HASH", build.rust_info.sha_short().unwrap_or(""))
+                       .replace("STAMP", build.rust_info.sha().unwrap_or(""));
```
Comment (alexcrichton) [inline src/bootstrap/doc.rs]: This all seems pretty familiar, perhaps copied from elsewhere in this file? Could that be deduplicated?

**Comment by alexcrichton** [comment]
lgtm modulo one minor comment.

But yeah if you're ok holding off this may be good to wait until after the branch

Also the travis failure looks like it may need attention

File: src/bootstrap/doc.rs (hunk 1 above)
Comment (steveklabnik) [inline src/bootstrap/doc.rs]: Yeah; I wasn't sure about it. There's just enough that's different....

Comment (alexcrichton) [inline src/bootstrap/doc.rs]: Let's at least refactor out what we can, especially these bits about version/short_hash/stamp/etc

**Comment by steveklabnik** [comment]
I've updated the book to be tidy-clean and re-enabled tidy for the book. Let's see what travis has to say...

**Comment by steveklabnik** [comment]
```
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1011)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1057)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 108)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1086)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1209)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 124)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1265)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1296)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1314)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 1330)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 184)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 345)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 377)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 387)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 401)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 505)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 525)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 542)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 635)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 658)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 723)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 742)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 865)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 891)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 90)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 901)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 917)
    C:\Users\steve\src\rust\src/doc\book\second-edition\nostarch\chapter15.md - Smart_Pointers (line 931)
```

well uh, okay then

**Comment by steveklabnik** [comment]
@alexcrichton I never know what the right way to compare paths is, please check out the latest commit there. 

This now passes locally on my machine, and so should be good to go sooN!

**Comment by steveklabnik** [comment]
No idea why tidy didn't see that on my machine...

**Comment by steveklabnik** [comment]
@bors: r=alexcrichton

I had to fix up a number of links because, silly me, I forgot to run linkchecker.

I am triple checking a clean build locally but there are a _lot_ of p=1 builds in the queue, so gonna r+ this now

**Comment by steveklabnik** [comment]
@bors: r=alexcrichton

embarrassingly, I rebased out an important linkchecker change. The clean build now passes on my system.

File: src/tools/linkchecker/main.rs [hunk 2]
```diff
@@ -121,6 +121,12 @@ fn check(cache: &mut Cache,
     if file.extension().and_then(|s| s.to_str()) == Some("js") {
         return None;
     }
+
```
Comment (frewsxcv) [inline src/tools/linkchecker/main.rs]: Tidy error: Trailing whitespace

Comment (steveklabnik) [inline src/tools/linkchecker/main.rs]: gah :(

**Comment by steveklabnik** [comment]
@bors: r-

okay i dunno why tidy is now flipping out :(

**Comment by steveklabnik** [comment]
@bors: r=alexcrichton

oh my directory was dirty. I need more ☕️ . Tidy passes.

**Comment by steveklabnik** [comment]
@bors: r=alexcrichton

passed everything on my machine

**Pull request merged**